include/range/v3/utility/move.hpp

```diff
@@ -66,10 +68,37 @@ namespace ranges
         /// \cond
         namespace adl_move_detail
         {
+
+            constexpr struct valid_expr_t
+            {
+                template<typename ...T>
+                void operator()(T &&...) const;
+            }  valid_expr {};
+
+            struct IndirectlyMovable_
+            {
+                template<typename I,
+                         typename R = decltype(*std::declval<I>()),
+                         typename U = meta::eval<std::remove_reference<R>>//,
+                >
+                auto requires_(I const& i) -> decltype(
+                    valid_expr(
+                        aux::move_t<R>{aux::move(*i)}
+                    ));
+            };
+            template<typename I>
+            using IndirectlyMovable
+            = concepts::models<IndirectlyMovable_, I>;
+
             // Default indirect_move overload.
             template<typename I,
-                typename R = decltype(*std::declval<I>()),
-                typename U = meta::eval<std::remove_reference<R>>>
+                  typename R = decltype(*std::declval<I>()),
+                  typename U = meta::eval<std::remove_reference<R>>,
+                  typename std::enable_if<IndirectlyMovable<I>{}, int>::type = 0        //,
```

Comment:
The only truly broken thing that I still have to fix is figuring out why GCC builds break if I constrain `indirect_move` in any way I've tried:

- the current approach with enable if (using IndirectlyMovable, I know I shouldn't do this)
- constraining directly by just adding a new type to the template parameter list that does `decltype(aux::move_t<R>(aux::move(*std::declval<I const&>())))`,
- constraining using expression SFINAE... (this was my preferred way since it is the simplest thing possible)

They all work with all versions of clang, and they all break with GCC. I am completely bluffed. I'll install gcc 4.9 and look into this again this week since trying to debug this through travis takes too much time.

There are some other things to discuss like the `inplace` method on algorithms that use `get_temporary_buffer` and if `nth_element` and `is_permutation` should be rewritten to make them constexpr but once the direction in these issues is clear implementing them should be easy.